### PR TITLE
Feature/enable reading parameters from payload in http action

### PR DESCRIPTION
**Pull request by marcinkp**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholdersResolverConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package io.knotx.server.common.placeholders;
```
Comment (malaskowski) [inline common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholdersResolverConfiguration.java:16]: Please add license headers.

File: common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholdersResolverConfigurationItem.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package io.knotx.server.common.placeholders;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+class PlaceholdersResolverConfigurationItem {
+
+  private final List<PlaceholderSubstitutor> placeholderSubstitutor;
+  private final List<String> prefixes;
```
Comment (malaskowski) [inline common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholdersResolverConfigurationItem.java]: Could be a `Set`, then you may implement `startsWith()` method as:
```
prefixes.contains(placeholder);
```

File: common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholderSubstitutor.java [hunk 3]
```diff
@@ -15,18 +15,18 @@
  */
 package io.knotx.server.common.placeholders;
 
-import io.knotx.server.api.context.ClientRequest;
+import java.util.List;
 
 @FunctionalInterface
-public interface PlaceholderSubstitutor {
+public interface PlaceholderSubstitutor<T> {
```
Comment (malaskowski) [inline common/placeholders/src/main/java/io/knotx/server/common/placeholders/PlaceholderSubstitutor.java:21]: Thank you for doing that 👍

File: common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import io.vertx.core.json.JsonObject;
+
+public class JsonPlaceholderSubstitutor implements PlaceholderSubstitutor<JsonObject> {
+
+  static final String PREFIX_PAYLOAD = "payload.";
+
+  @Override
+  public String getValue(JsonObject source, String placeholder) {
```
Comment (malaskowski) [inline common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java:29]: There could be some unit tests for this class.

File: common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+import static io.knotx.commons.json.JsonObjectUtil.getString;
+
+import org.apache.commons.lang3.StringUtils;
+
+import io.vertx.core.json.JsonObject;
+
+public class JsonPlaceholderSubstitutor implements PlaceholderSubstitutor<JsonObject> {
+
+  static final String PREFIX_PAYLOAD = "payload.";
```
Comment (malaskowski) [inline common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java]: I wonder if this is proper prefix...
Knot.x Server does not know anything about Knot.x Fragments, and `payload` is part of a fragment.
I'm also not 100% satisfied with prefix like `json.` - maybe you have any better idea?

Comment (marcinkp) [reply common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java]: not really. I assumed that payload is generic enough :)

Comment (malaskowski) [reply common/placeholders/src/main/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutor.java]: Ok, let's leave it for now.

**Review by malaskowski — APPROVED** [review]

File: common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  @Test
+  public void shouldReturnValueForBoolean() {
+    //given
+    JsonPlaceholderSubstitutor underTest = new JsonPlaceholderSubstitutor();
+
+    //when
+    String booleanValue = underTest.getValue(payload, "payload.values.boolean");
+
+    //then
+    assertEquals("true", booleanValue);
```
Comment (tomaszmichalak) [inline common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java:64]: Why not boolean here?

File: common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  @Test
+  public void shouldReturnValueForNumber() {
+    //given
+    JsonPlaceholderSubstitutor underTest = new JsonPlaceholderSubstitutor();
+
+    //when
+    String numberValue = underTest.getValue(payload, "payload.values.number");
+
+    //then
+    assertEquals("1", numberValue);
```
Comment (tomaszmichalak) [inline common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java:52]: Please use Integer here.

File: common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java (hunk 6 above)
Comment (marcinkp) [reply common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java:64]: as interface method returns String. and placeholder is part of parametrized String...

File: common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java (hunk 7 above)
Comment (marcinkp) [reply common/placeholders/src/test/java/io/knotx/server/common/placeholders/JsonPlaceholderSubstitutorTest.java:52]: as above.

**Pull request merged**